upload/src/addons/SV/WarningImprovements/XF/Service/Conversation/Notifier.php

```diff
@@ -4,31 +4,43 @@
 
 class Notifier extends XFCP_Notifier
 {
+    protected $sv_force_email_for_user_id = null;
+    protected $sv_respect_receive_admin_email = true;
+
     protected function _canUserReceiveNotification(\XF\Entity\User $user, \XF\Entity\User $sender = null)
     {
         if (!empty(\SV\WarningImprovements\Listener::$warnngObj))
         {
             if (\SV\WarningImprovements\Listener::$warnngObj->user_id == $user->user_id)
             {
-                $options = $this->app->options();
-
-                $this->conversation->setOption('warningObj', \SV\WarningImprovements\Listener::$warnngObj);
-
-                if ($options->sv_force_conversation_email_on_warning)
+                if ($this->sv_force_email_for_user_id === null)
                 {
-                    $receiveAdminEmail = $user->Option->receive_admin_email;
+                    $this->sv_force_email_for_user_id = 0;
+                    $options = $this->app->options();
 
-                    if (!$options->sv_respect_receive_admin_email_on_warning)
+                    if ($options->sv_force_conversation_email_on_warning)
                     {
-                        $receiveAdminEmail = true;
+                        $this->sv_respect_receive_admin_email = $options->sv_respect_receive_admin_email_on_warning;
+                        $this->sv_force_email_for_user_id = \SV\WarningImprovements\Listener::$warnngObj->user_id;
                     }
-
                     if ($options->sv_only_force_warning_email_on_banned && !$user->is_banned)
                     {
-                        $receiveAdminEmail = false;
+                        $this->sv_force_email_for_user_id = 0;
+                    }
+                }
+
+                if ($this->sv_force_email_for_user_id)
+                {
+                    if ($this->sv_respect_receive_admin_email)
+                    {
+                        $user->Option->email_on_conversation = $user->Option->receive_admin_email;
```

Comment:
I'ld prefer to not make changes to a user's entity object which might somehow get accidentally saved. Especially as an entity is global to the request.

My preferred solution would be to override `_canUserReceiveNotification()` and if it returns false (and onlyNotifyUsers is empty or not matching), then force it.